SITL: Add support for AirSim Simulator

rajat2004 opened the pull request:
Connector for AirSim based on the ArduCopter Solo API in AirSim: https://github.com/microsoft/AirSim/pull/1387
Described properly with some relevant links in #11324
A WIP, need to figure out some connection issues.

Review Requested @tridge @peterbarker

rajat2004 commented:
@OXINARF Thanks!

The connection problem is described in detail here - https://github.com/microsoft/AirSim/issues/1965
Briefly - Ardupilot is binding the sockets correctly and Airsim is also connecting to the sockets. Airsim is attempting to send the data but the SITL isn't receiving anything

tridge commented:
@rajat2004 I've cleaned up the commits using an interactive rebase to follow our commit style 
See http://ardupilot.org/dev/docs/submitting-patches-back-to-master.html

lucasdemarchi commented on libraries/SITL/SIM_AirSim.cpp:
```diff
@@ -0,0 +1,158 @@
+	uint64_t *buf = (uint64_t *)this;
+	for (uint16_t i=0; i<sizeof(*this)/8; ++i) {
+		buf[i] = le64toh(buf[i]);
+	}
```
this doesn't make much sense - fdm_packet should be a pod, and if you need to convert endiannes you do it outside.

Also, unless you are using a powerpc, sparc or some flavors of ARM, your host cpu is also little endian, so this is a nop.

peterbarker replied:
This also won't be idempotent where swapping is done.  IOW doing this in place is a little odd.

rajat2004 replied:
I have removed this function for now since as @lucasdemarchi pointed out, this would be a nop for most of the systems. Also the struct is packed so need to check with that also

rajat2004 commented:
@tridge I thought that I would squash up the commits after the connection problems were fixed, but thanks for doing that

rajat2004 commented:
@lucasdemarchi Sorry, pressed the resolved button by mistake. If possible, could you open it again
The conversion from little-endian was something which I added after seeing a similar thing in SIM_JSBSim
https://github.com/ArduPilot/ardupilot/blob/869a369cc007c39a7cf1717859fa2483e3a13432/libraries/SITL/SIM_JSBSim.cpp#L391

Even I thought that using little-endian wouldn't be much of a problem, but added that just for making sure.
Should I remove the function?

rajat2004 commented:
@lucasdemarchi Sorry, pressed the resolved button by mistake. If possible, could you open it again
The conversion from little-endian was something which I added after seeing a similar thing in SIM_JSBSim
https://github.com/ArduPilot/ardupilot/blob/869a369cc007c39a7cf1717859fa2483e3a13432/libraries/SITL/SIM_JSBSim.cpp#L391

Even I thought that using little-endian wouldn't be much of a problem, but added that just for making sure.
Should I remove the function?

peterbarker commented on libraries/SITL/SIM_AirSim.cpp:
```diff
@@ -0,0 +1,158 @@
+	Aircraft(home_str, frame_str),
+	sensor_sock(true),
+	control_sock(true),
+	last_timestamp(0),
```
No need to zero these; we will never make one of these on the stack and elsewhere it gets zeroed.

peterbarker commented on libraries/SITL/SIM_AirSim.cpp:
```diff
@@ -0,0 +1,158 @@
+
+void AirSim::send_servos(const struct sitl_input &input)
+{
+	servo_packet pkt;
```
This should be brace-initialised to zero.

rajat2004 replied:
Done

peterbarker commented on libraries/SITL/SIM_AirSim.cpp:
```diff
@@ -0,0 +1,158 @@
+{
+	servo_packet pkt;
+
+	for(int i=0; i<kArduCopterRotorControlCount; ++i) {
```
We use `uint8_t` for loop counters in ardupilot and always post-increment rather than pre-increment.  That's a style thing (and the google style guide says the opposite... :-) )

rajat2004 replied:
Done

rajat2004 replied:
Should this be added to the style-guide?

peterbarker commented on libraries/SITL/SIM_AirSim.cpp:
```diff
@@ -0,0 +1,158 @@
+	pkt.direction = input.wind.direction;
+	pkt.turbulance = input.wind.turbulence;
+
+	control_sock.send(&pkt, sizeof(pkt));
```
Need to check this return value.

rajat2004 replied:
Done

peterbarker commented on libraries/SITL/SIM_AirSim.cpp:
```diff
@@ -0,0 +1,158 @@
+}
+
+/*
+	Recieve new sensor data from simulator
```
`s/Recieve/Receive/`

rajat2004 replied:
Fixed

peterbarker commented on libraries/SITL/SIM_AirSim.cpp:
```diff
@@ -0,0 +1,158 @@
+*/
+void AirSim::recv_fdm(const struct sitl_input &input)
+{
+	fdm_packet pkt;
```
Move this to just above the loop (below the comments)

rajat2004 replied:
Done

peterbarker commented on libraries/SITL/SIM_AirSim.cpp:
```diff
@@ -0,0 +1,158 @@
+{
+	fdm_packet pkt;
+
+	// Resend servo packet every 0.1 second till we get data
```
Why is this looping required?

rajat2004 replied:
From my understanding, this allows some packet loss while receiving the fdm. This has been done in most of the other simulator connectors also

peterbarker commented on libraries/SITL/SIM_AirSim.cpp:
```diff
@@ -0,0 +1,158 @@
+	}
+	// NOTE: Airsim sends data in little-endian
+
+	// Taken from CRRCSim.cpp
```
Does AirSim conform to NED?

rajat2004 replied:
Need to confirm that. I didn't fix all this till now since I was focusing on making sure that they are able to communicate first and then I could proceed ahead

rajat2004 replied:
https://github.com/microsoft/AirSim/blob/master/docs/apis.md#coordinate-system
Internally, no. Fixed that while calculating position

peterbarker commented on libraries/SITL/SIM_AirSim.cpp:
```diff
@@ -0,0 +1,158 @@
+    gyro = Vector3f(pkt.rollRate, pkt.pitchRate, pkt.yawRate);
+    velocity_ef = Vector3f(pkt.speedN, pkt.speedE, pkt.speedD);
+
+    Location loc1, loc2;
```
Use brace-initilisation to create these Locations and create them as const.

rajat2004 replied:
Has been removed entirely

peterbarker commented on libraries/SITL/SIM_AirSim.h:
```diff
@@ -0,0 +1,86 @@
+	static const int kArduCopterRotorControlCount = 11;
+
+	struct servo_packet {
+		// ArduPilot Solo rotor control datagram format
```
We should probably not reference Solo in here unless stuff really is Solo-specific :-)

rajat2004 replied:
Well, yes, this entire connector is following the old Solo format since, in Airsim, they have added support for that

rajat2004 replied:
Removed

peterbarker commented on libraries/SITL/SIM_AirSim.cpp:
```diff
@@ -0,0 +1,158 @@
+
+	// Resend servo packet every 0.1 second till we get data
+	// We do get the timestamp in the packet, sync can be achieved in a similar manner as JSBSim
+	while (sensor_sock.recv(&pkt, sizeof(pkt), 100) != sizeof(pkt)) {
```
Error checking?

rajat2004 replied:
Done

peterbarker commented on libraries/SITL/SIM_AirSim.cpp:
```diff
@@ -0,0 +1,158 @@
+		return true;
+	}
+	// binding to 0.0.0.0 or 127.0.0.1 ?
+	if(!control_sock.bind(airsim_ip, airsim_control_port)) {
```
Minor code-formatting issue here.  You might consider running `astyle` to avoid a flood of similar comments :-)

rajat2004 replied:
Formatting fixed

rajat2004 commented on libraries/SITL/SIM_AirSim.cpp:
```diff
@@ -0,0 +1,158 @@
+	if (opened_sensor_socket) {
+		return true;
+	}
+	// This should be binding to 127.0.0.1 or 0.0.0.0 ?
```
Fixed

rajat2004 commented on libraries/SITL/SIM_AirSim.cpp:
```diff
@@ -0,0 +1,158 @@
+	}
+	// This should be binding to 127.0.0.1 or 0.0.0.0 ?
+	if(!sensor_sock.bind(airsim_ip, airsim_sensor_port)) {
+		printf("Unable to open Airsim sensor socket at %s:%u\n", airsim_ip, airsim_sensor_port);
```
Done

rajat2004 commented on libraries/SITL/SIM_AirSim.cpp:
```diff
@@ -0,0 +1,158 @@
+	if (opened_control_socket) {
+		return true;
+	}
+	// binding to 0.0.0.0 or 127.0.0.1 ?
```
Removed

rajat2004 commented on libraries/SITL/SIM_AirSim.cpp:
```diff
@@ -0,0 +1,158 @@
+/*
+	Connect to required sockets
+*/
+bool AirSim::connect_sensor_socket(void)
```
Fixed

rajat2004 commented:
Fixed the connection issues. 
Please note that this is in no way functional right now, the SITL isn't even initialising properly. 
I'll add a readme also for how to build AirSim and use this Ardupilot connector

rajat2004 commented:
Readme added

rajat2004 commented on libraries/SITL/SIM_AirSim.cpp:
```diff
@@ -0,0 +1,158 @@
+{
+	servo_packet pkt;
+
+	for(int i=0; i<kArduCopterRotorControlCount; ++i) {
```
Already fixed

rajat2004 commented on libraries/SITL/SIM_AirSim.cpp:
```diff
@@ -0,0 +1,158 @@
+	if (opened_sensor_socket) {
+		return true;
+	}
+	// This should be binding to 127.0.0.1 or 0.0.0.0 ?
```
Forgot to change that, thanks for pointing out

rajat2004 commented:
Fixed the gyro initialization by using EKF 10. The bad values are probably due to the collision of the drone with the ground in the beginning. Still figuring out the GPS problem

rajat2004 commented on libraries/SITL/SIM_AirSim.cpp:
```diff
@@ -0,0 +1,151 @@
+
+    // position.x = pkt.latitude;
+    // position.y = pkt.longitude;
+    // position.z = pkt.altitude;
```
Using this instead of the above block of code gives the correct GPS, but this seems incorrect since in the description in SIM_Aircraft, it's written `meters, NED from origin`

rajat2004 commented on libraries/SITL/SIM_AirSim.cpp:
```diff
@@ -0,0 +1,151 @@
+    Location loc1, loc2;
+    loc2.lat = pkt.latitude * 1.0e7;
+    loc2.lng = pkt.longitude * 1.0e7;
+    const Vector2f posdelta = loc1.get_distance_NE(loc2);
```
Is this proper? Shouldn't the distance be taken from home rather than from 0

peterbarker commented:
On Tue, 21 May 2019, Rajat Singhal wrote:

> +	for(int i=0; i<kArduCopterRotorControlCount; ++i) {
> 
> Should this be added to the style-guide?

Probably :-)

rajat2004 commented:
Made some changes, now it doesn't get the random GPS location. But still, there's no sat fix. It's not even detecting the GPS as u-blox which is displayed in the Mavproxy console. Any idea on what is causing this?
The GPS sent by the simulator is consistent (cause the drone won't take off)

I haven't squashed this commit so that it would be easier to review these changes and rewind if required

rajat2004 commented:
Also seeing `Time has wrapped` messages now. Probably due to incorrect use of the `timestamp` field.
Or because of the horrible timing between Ardupilot & Airsim right now?

rajat2004 commented:
Yeah, so the problem is with the timestamp, AirSim sends time from 1970 rather than the simulator start

rajat2004 commented:
Did a bit of hack, but it's working!!
Getting the GPS lock also, able to arm
While trying to takeoff, the rotors are spinning but it doesn't move, it detects it as crash and then disarms.

But still, progress!
I'll push the changes in a few min after a bit of cleanup

rajat2004 commented:
Added the sensor conversion changes.
Also added the params file, @tridge Thanks a lot for that!
Right now it's the full list, will change it to only those which have been modified
Readme has also been updated

rajat2004 commented:
Rebased on master, removed the redundant params, flying quite nicely
@peterbarker @tridge Should I squash the commits now?

khancyr commented on Tools/autotest/pysim/vehicleinfo.py:
```diff
@@ -92,6 +92,10 @@ def __init__(self):
                 "default_params_filename": ["default_params/copter.parm",
                                             "default_params/gazebo-iris.parm"],
             },
+            "airsim": {
```
airsim-copter? to allow other vehicle in future

rajat2004 replied:
Didn't notice that, I'll change it

rajat2004 replied:
Done

rajat2004 replied:
Once we have hexa configuration in Airsim, then this can be changed to airsim-quadx

khancyr commented on libraries/SITL/SIM_AirSim.cpp:
```diff
@@ -0,0 +1,165 @@
+		return true;
+	}
+
+	if (!sock.bind("0.0.0.0", airsim_sensor_port)) {
```
why not use set_interface_ports() here ? it will allow you to select the port you want

rajat2004 replied:
That wasn't used in the other backends, so didn't use it in this one also.
Now checked the Gazebo files, will use it
Thanks for the review!!

rajat2004 replied:
I can make this change but then this will set the ports to 9002 & 9003 which are different from the default ports used by Airsim. The ports can be set through `settings.json`

Just confirming once before making the changes

rajat2004 replied:
Done

rajat2004 commented:
This should now work on different machines
Things to change for running on different machines & using different ports have been mentioned in the Readme, but haven't been tested

rajat2004 commented:
I have removed the last commit which changed the frame_rate based on the difference between the last timestamp received and the current packet's. Therefore now it's back to the fixed 3000us timestep

This was done since first, Airsim was sending the wall clock time rather than the stepped time. This had variation from the timestep being used by the Steppable timer used internally by Airsim, and so caused the frame_time to updated wrongly.

I also modified the ArduCopter Solo API slightly to send the Stepped simulation time instead of the wall-clock time (https://github.com/rajat2004/AirSim/commit/205515e0eb67408b1673f2f8b9844746a54cdcb3)
Logically after this, the variable timestep should work
And it did work, however what I found is that even this is not foolproof, sometimes Airsim does multiple steps, due to this, the frame_time would again be updated with a worng value. But all other times, the difference was exactly 3000us
Below, is some of the timestamp data collected which shows the incorrect timejumps

```
Deltat: 6001, TimeNow: 20563272, Timestamp: 1560414075807410 
Deltat: 6000, TimeNow: 20692274, Timestamp: 1560414075936412 
Deltat: 6000, TimeNow: 20947280, Timestamp: 1560414076191418 
Deltat: 246006, TimeNow: 21316288, Timestamp: 1560414076560426 
Deltat: 9000, TimeNow: 21700296, Timestamp: 1560414076944434 
Deltat: 42001, TimeNow: 21865299, Timestamp: 1560414077109437 
Deltat: 117003, TimeNow: 23995345, Timestamp: 1560414079239483 
Deltat: 117003, TimeNow: 26629401, Timestamp: 1560414081873539 
Deltat: 6000, TimeNow: 27136412, Timestamp: 1560414082380550 
Deltat: 105002, TimeNow: 29254457, Timestamp: 1560414084498595 
Deltat: 114002, TimeNow: 31885513, Timestamp: 1560414087129651 
Deltat: 6000, TimeNow: 32770532, Timestamp: 1560414088014670 
Deltat: 6000, TimeNow: 32899535, Timestamp: 1560414088143673 
Deltat: 111002, TimeNow: 34519569, Timestamp: 1560414089763707 
Deltat: 6000, TimeNow: 35152583, Timestamp: 1560414090396721 
Deltat: 6000, TimeNow: 35407588, Timestamp: 1560414090651726 
Deltat: 6000, TimeNow: 35914599, Timestamp: 1560414091158737
```

rajat2004 commented:
This PR is no longer compatible with the ArduCopterSoloAPI in Airsim, instead, the corresponding Airsim branch is https://github.com/rajat2004/AirSim/tree/pr-arducopter.
All further development will be done on this branch

The earlier Airsim Solo compatible branch is here - https://github.com/rajat2004/ardupilot/tree/airsim-soloapi

rajat2004 commented:
JSON sensor packet parsing has been added, corresponding changes have been made to the Airsim branch

rajat2004 commented:
@peterbarker @tridge @khancyr @lucasdemarchi Could you do a fresh review of the PR as it currently stands? All the points raised till now have been addressed

rajat2004 commented:
@tridge Thanks!
Should we wait for the AirSim PR to get merged first? (however, that might take a long time, and merging this will make it easier to setup and for other developers to work on top of this)

Just wanted to confirm once

tridge commented:
@rajat2004 no, we should merge here first I think. done!
I know you will be adding more features, but that can be done in a followup PR

rajat2004 commented:
Follow-up PR for adding Lidar is here - https://github.com/ArduPilot/ardupilot/pull/11835
Also, https://github.com/ArduPilot/ardupilot/pull/11821 is a PR for fixing ArduPilot build on MacOS

Please have a look at the above PRs and review
